### PR TITLE
Forward headers (PR 1 of 3): The actual feature

### PR DESCRIPTION
**Pull request by rdrongow**
Hey! I hope i got it right now. I was struggeling with the call_args representation so I choose just to compare the strings. It's pretty much the same though. The new test is in line 224.

### CODE REVIEW

**Comment by rdrongow** [comment]
## BTW: This is a small pyramid app I wrote to actually see the whole thing in a brwose

  from waitress import serve
  from pyramid.config import Configurator
  from pyramid.response import Response
  from wesgi import MiddleWare
  from wesgi import Policy
  import pprint

  def esi_test(request):
    result = """
        <pre>
         Actual Request:

```
           {}


     ESI-Include Request:

     <esi:include src="http://localhost:8080/request_print" alt="URI" onerror="continue" />
    </pre>
""".format(pprint.pformat(dict(request.headers.items())))

return Response(result)
```

  def request_print(request):

```
return pprint.pformat(dict(request.headers.items()))
```

  if **name** == '**main**':
    config = Configurator()
    config.add_route('esi', '/esi')
    config.add_view(esi_test, route_name='esi')

```
config.add_route('request_print', '/request_print')
config.add_view(
    request_print, route_name='request_print', renderer='string')
esiapp = config.make_wsgi_app()
app = MiddleWare(
    esiapp, debug=False, forward_headers=True, policy=Policy())

serve(app, host='0.0.0.0', port=8080)
```

File: wesgi/__init__.py [hunk 1]
```diff
@@ -144,27 +146,32 @@ def delete(key):
...
         if resp.content_type == 'text/html' and resp.status_int == 200:
             orig_scheme = environ['wsgi.url_scheme']
-            new_body = self._process(resp.body, orig_scheme)
+            new_body = self._process(resp.body, orig_scheme, headers=headers)
             if new_body is not None:
                 resp.body = new_body
         return resp(environ, start_response)
 
-    def _process(self, body, orig_scheme):
+    def _process(self, body, orig_scheme, headers={}):
```
Comment (jinty) [inline wesgi/__init__.py:172]: please don't use {} as the default of a keyword argument. That's a bug magnet as the dict is only instantiated once.

Please do something like this:

```
def _process(self, body, orig_scheme, headers=None):
    if headers is None:
        headers = {}
```

I'm probably missing something, but I don't see why you need to use a keyword argument over a positional argument here.

File: wesgi/tests.py [hunk 2]
```diff
@@ -21,7 +20,7 @@ def mock_http_request(http, response=None, content=None):
             response = Response()
         mock.return_value = response, content
 
-def Response(status=200, headers=None):
+def Response(status=200, headers={}):
```
Comment (jinty) [inline wesgi/tests.py:23]: same "mutable used as keyword default" bug here

File: wesgi/tests.py [hunk 3]
```diff
@@ -31,7 +30,11 @@ def Response(status=200, headers=None):
 
 def run_mw(mw):
     start_response = Mock()
+
     request = webob.Request.blank("")
+    if mw.policy.forward_headers:
+        request.headers['Content-Type'] = 'text/xml'
+
```
Comment (jinty) [inline wesgi/tests.py:37]: could you rather pass in the request headers as a keyword argument to run_mw()? That way the test decides what headers to pass.

File: wesgi/tests.py [hunk 4]
```diff
@@ -99,7 +106,7 @@ def side_effect(*args):
         data = mw._process_include('level-1<esi:include src="http://www.example.com"/>-after')
         self.assertEquals(data, 'level-1-2-3-4-5-6-7-8-9-10-11-last-after')
         self.assertEquals(mock.call_count, 11)
-        self.assertEquals(mock.call_args, (('http://www.example.com/10', ), ))
+        self.assertEquals(str(mock.call_args), "call('http://www.example.com/10', headers={})")
```
Comment (jinty) [inline wesgi/tests.py:109]: hmm, I didn't know you could call str() on call_args. looks easier, but I think I've been too burnt with doctests in the past to like it...

just a the gripes of an old man, please don't change the PR for this

**Comment by jinty** [comment]
looking good, almost there, I just found 2 things which I think need changing before I merge.

**Comment by jinty** [comment]
equivalent functionality merged in https://github.com/jinty/wesgi/pull/5

**Pull request closed without merging**